Add property binding in torchbind

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50670 Add property binding in torchbind**

Summary: This PR adds property support to Torchbind. To handle getter and setter functions for custom class properties, we introduce __setattr__ method for ScriptObject that directly calls setter method on the C++ object. 

Test Plan: test cases in test_torchbind.py

Reviewers: @gmagogsfm, @SplitInfinity 

Subscribers:

Tasks:

Tags:

Differential Revision: [D26632781](https://our.internmc.facebook.com/intern/diff/D26632781)

tugsbayasgalan committed:
[WIP] add property binding in torchbind

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] add property binding in torchbind"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan commented:
This is just a early prototype for Torchbind property

SplitInfinity commented on test/cpp/jit/test_custom_class_registrations.cpp:
```diff
@@ -33,6 +33,27 @@ struct Foo : torch::CustomClassHolder {
+  FooGetterSetter() : x(0), y(0) {}
+  FooGetterSetter(int64_t x_, int64_t y_) : x(x_), y(y_) {}
+
+  int64_t getX() {
+    return x;
+  }
+  void setX(int64_t z) {
+    x = z;
+  }
+
+  int64_t getY() {
+    return y;
+  }
```
I would make these more complicated just to ensure that the tests aren't passing because of the getters actually being attribute lookups.

SplitInfinity commented on torch/custom_class.h:
```diff
@@ -153,6 +153,26 @@ class class_ {
+        detail::wrap_func<CurClass, GetterFunc>(std::move(getterFunc));
+    auto wrapped_setter =
+        detail::wrap_func<CurClass, SetterFunc>(std::move(setterFunc));
+    defineProperty(
```
Unless you have plans to reuse this function, you can include the body of `defineProperty` here.

tugsbayasgalan replied:
sounds good

SplitInfinity commented on torch/custom_class.h:
```diff
@@ -153,6 +153,26 @@ class class_ {
+      std::string name,
+      GetterFunc getterFunc,
+      SetterFunc setterFunc,
+      std::string getter_doc_string = "",
+      std::string setter_doc_string = "") {
```
I think the user experience would be better with only one docstring (for the property) instead of two.

tugsbayasgalan replied:
sounds good

SplitInfinity reviewed (commented):
Nice!

tugsbayasgalan commented on test/cpp/jit/test_custom_class_registrations.cpp:
```diff
@@ -210,6 +231,10 @@ TORCH_LIBRARY(_TorchScriptTesting, m) {
 
+  m.class_<FooGetterSetter>("_FooGetterSetter")
+      .def(torch::init<int64_t, int64_t>())
+      .def_property("x", &FooGetterSetter::getX, &FooGetterSetter::setX);
```
Yep will add it :)

tugsbayasgalan committed:
Update on "[WIP] add property binding in torchbind"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] add property binding in torchbind"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] add property binding in torchbind"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] add property binding in torchbind"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] add property binding in torchbind"


Summary: This PR adds property support to Torchbind. Inside Torchscript, it

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan commented:
@SplitInfinity @gmagogsfm could you guys take a look?

tugsbayasgalan committed:
Update on "[WIP] add property binding in torchbind"


Summary: This PR adds property support to Torchbind. There are two cases that it needs to work:

**Torchscript**
Inside Torchscript, we don't go through pybind so there is no issue with accessing properties through ClassType.

**Eager Mode**
In Eager Mode, Torchbind creates ScriptObject which we cannot dynamically add (aka access) properties after initializing it. (https://stackoverflow.com/questions/1325673/how-to-add-property-to-a-class-dynamically
) Therefore we created a Python wrapper (RecursiveScriptClass) around ScriptObject where we can use property method to set properties.  By doing so, we can look up wrapped object's property through __getattr__ method of the RecursiveScriptClass. This logic is inspired from https://github.com/pytorch/pytorch/pull/44324

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

gmagogsfm commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1881,8 +1881,8 @@ struct TORCH_API ClassType : public NamedType {
   // getter and (optional) setter for that attribute.
   struct Property {
     std::string name;
-    const torch::jit::Function* getter;
```
I am sure this is necessary change, but I didn't find code that modifies getter/setter. Could you point me to them?

tugsbayasgalan replied:
These are used (https://github.com/pytorch/pytorch/pull/50670/files#diff-c699f539322d99f6ab85853a11d9b7339acb49639d8797934c8fec2faa5ca7d4R109 AND https://github.com/pytorch/pytorch/pull/50670/files#diff-c699f539322d99f6ab85853a11d9b7339acb49639d8797934c8fec2faa5ca7d4R112)

gmagogsfm commented on torch/csrc/jit/python/module_python.h:
```diff
@@ -16,5 +16,13 @@ inline c10::optional<Module> as_module(const py::object& obj) {
   return c10::nullopt;
 }
 
+inline c10::optional<Object> as_object(const py::object& obj) {
```
Where is this used?

tugsbayasgalan replied:
deleted

gmagogsfm commented on test/cpp/jit/test_custom_class_registrations.cpp:
```diff
@@ -210,6 +246,16 @@ TORCH_LIBRARY(_TorchScriptTesting, m) {
       .def("add", &Foo::add)
       .def("combine", &Foo::combine);
 
+  m.class_<FooGetterSetter>("_FooGetterSetter")
```
I think using lambda's as getter and setter already works with your implementation, could we add test cases covering lambdas? Lambdas are really important because TorchScript only supports a subset of C++ types, users frequently rely on lambdas to cast types back-and-forth between TorchScript friendly types and actual C++ types.

tugsbayasgalan replied:
added a new test.

gmagogsfm commented on test/jit/test_torchbind.py:
```diff
@@ -130,6 +130,47 @@ def foo():
         self.assertEqual(out[1].pop(), "hi")
 
+    def test_torchbind_def_property(self):
+        def foo_getter_setter_full():
```
Nit:

Consider splitting `test_torchbind_def_property` into three test cases, so that they can be better parallelized in testing.

tugsbayasgalan replied:
yep will do on the final commit.

gmagogsfm commented on torch/_jit_internal.py:
```diff
@@ -658,7 +658,7 @@ def get_static_fn(cls, fn):
 
 def get_torchscript_modifier(fn):
-    if not callable(fn):
+    if not callable(fn) or isinstance(fn, torch.jit.RecursiveScriptClass):
```
I am probably missing something, but why is this special case needed?

Based on my naive understanding, if we remove this special case, `get_torchscript_modifier` called on `torch.jit.RecursiveScriptClass` instances would return `FunctionModifiers.DEFAULT`. Then TorchScript would attempt to script the instance, which would hit line 1083 of `_script.py`. And the final result would be correct. Is this accurate or am I missing something obvious?

tugsbayasgalan replied:
yep deleted

gmagogsfm reviewed (commented):
Looks awesome! Just have some minor comments and question for my own understanding. Thanks for making this happen!

tugsbayasgalan commented on torch/custom_class.h:
```diff
@@ -153,6 +153,79 @@ class class_ {
+    getter = process_and_register_property_method(
+        getterName, wrapped_getter, doc_string);
+
+    if (!std::is_same<SetterFunc, GetterFunc>::value) {
```
Yeah. The compiler was complaining about nullptr_t. So this is used as a way to figure out if there is any setter method given.

SplitInfinity replied:
I think we talked about this long ago but can you remind us what the issue is with using `nullptr_t` as the default value for `SetterFunc`? I don't think `SetterFunc = GetterFunc` and using that to infer whether there is a getter or setter is a clean solution, even though a getter and setter can't have the same type.

tugsbayasgalan replied:
This method internally calls this (https://github.com/pytorch/pytorch/blob/22a34bcf4e5eaa348f0117c414c3dd760ec64b13/c10/util/TypeTraits.h#L79) But issue is that compiler errors because it thinks std::nullptr_t doesn't have operator(). Unless our SetterFunc has a default type that defines operator(), I don't think it will work.

tugsbayasgalan committed:
Update on "[WIP] add property binding in torchbind"


Summary: This PR adds property support to Torchbind. There are two cases that it needs to work:

**Torchscript**
Inside Torchscript, we don't go through pybind so there is no issue with accessing properties through ClassType.

**Eager Mode**
In Eager Mode, Torchbind creates ScriptObject which we cannot dynamically add (aka access) properties after initializing it. (https://stackoverflow.com/questions/1325673/how-to-add-property-to-a-class-dynamically
) Therefore we created a Python wrapper (RecursiveScriptClass) around ScriptObject where we can use property method to set properties.  By doing so, we can look up wrapped object's property through __getattr__ method of the RecursiveScriptClass. This logic is inspired from https://github.com/pytorch/pytorch/pull/44324

Test Plan: test cases in test_torchbind.py

Reviewers: @gmagogsfm, @SplitInfinity 

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] add property binding in torchbind"


Summary: This PR adds property support to Torchbind. There are two cases that it needs to work:

**Torchscript**
Inside Torchscript, we don't go through pybind so there is no issue with accessing properties through ClassType.

**Eager Mode**
In Eager Mode, Torchbind creates ScriptObject which we cannot dynamically add (aka access) properties after initializing it. (https://stackoverflow.com/questions/1325673/how-to-add-property-to-a-class-dynamically
) Therefore we created a Python wrapper (RecursiveScriptClass) around ScriptObject where we can use property method to set properties.  By doing so, we can look up wrapped object's property through __getattr__ method of the RecursiveScriptClass. This logic is inspired from https://github.com/pytorch/pytorch/pull/44324

Test Plan: test cases in test_torchbind.py

Reviewers: @gmagogsfm, @SplitInfinity 

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add property binding in torchbind"


Summary: This PR adds property support to Torchbind. There are two cases that it needs to work:

**Torchscript**
Inside Torchscript, we don't go through pybind so there is no issue with accessing properties through ClassType.

**Eager Mode**
In Eager Mode, Torchbind creates ScriptObject which we cannot dynamically add (aka access) properties after initializing it. (https://stackoverflow.com/questions/1325673/how-to-add-property-to-a-class-dynamically
) Therefore we created a Python wrapper (RecursiveScriptClass) around ScriptObject where we can use property method to set properties.  By doing so, we can look up wrapped object's property through __getattr__ method of the RecursiveScriptClass. This logic is inspired from https://github.com/pytorch/pytorch/pull/44324

Test Plan: test cases in test_torchbind.py

Reviewers: @gmagogsfm, @SplitInfinity 

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan commented on torch/jit/_script.py:
```diff
@@ -270,6 +269,154 @@ def __getattr__(self, attr):
+        # __setstate__ method to its' cpp class. Therefore, this reduce method manually calls init inside
+        # __setstate__ so that we can avoid infinite recursion. The solution was suggested here:
+        # https://stackoverflow.com/questions/50308214/python-3-alternatives-for-getinitargs
+        def __reduce__(self):
```
this is bit weird.

SplitInfinity replied:
Can you explain the need for this further?

tugsbayasgalan replied:
Basically when we call **__setstate__** in python, it doesn't actually cal inside **__init__** inside so it infinetely recurses inside **__getattr__** because we don't have the properties inside. Therefore, we can use **__reduce__** to call the constructor so that the props are set.

tugsbayasgalan commented on torch/jit/_trace.py:
```diff
@@ -1038,6 +1037,8 @@ def check_unique(param):
                 check_unique(buf)
         for name, val in orig.__dict__.items():
+            if isinstance(val, ScriptObjectWrapper):
+                val = val._c
```
is this okay?

tugsbayasgalan commented on test/quantization/test_quantized_op.py:
```diff
@@ -3019,8 +3019,8 @@ def _test_embedding_bag_unpack_fn(self, pack_fn, unpack_fn, num_embeddings, embe
-            self.assertEqual(isinstance(real_packed_weight, torch._C.ScriptObject), True)
-            unpacked_weight = torch.ops.quantized.embedding_bag_unpack(real_packed_weight)
+            self.assertEqual(isinstance(real_packed_weight._c, torch._C.ScriptObject), True)
+            unpacked_weight = torch.ops.quantized.embedding_bag_unpack(real_packed_weight._c)
```
@SplitInfinity just mentioning cuz quantization issue we talked about

SplitInfinity commented on torch/custom_class.h:
```diff
@@ -177,6 +177,80 @@ class class_ {
+  template <typename GetterFunc, typename SetterFunc = GetterFunc>
+  class_& def_property(
+      std::string name,
+      GetterFunc getter_func,
+      SetterFunc setter_func = nullptr,
```
We should probably have some check that makes sure `getter_func` and `setter_func` have reasonable schemas (e.g. getter must return something, setter cannot return anything).

SplitInfinity commented on torch/jit/_script.py:
```diff
@@ -270,6 +269,154 @@ def __getattr__(self, attr):
+        def __setstate__(self, val):
+            self.forward_magic_method("__setstate__", val)
+
+        def __iter__(self):
```
We probably don't need all of these `__` methods for Torchbind. I think we will expand this list later on, but for now I think we should be able to get by with just `__getstate__` and `__setstate__` and maybe a few others.

tugsbayasgalan replied:
ok

SplitInfinity reviewed (commented):
I think the basics are correct, but some cleanup is required.

In addition, the introduction of `ScriptObjectWrapper` is a lot messier than I thought it would be and so I have two questions:
1. Do we really need to enable using properties on TorchBind objects outside of TorchScript (cc: @gmagogsfm) ? As I see it, Torchbind is a solution for FFI between TorchScript and C++, not Python and C++ (that's pybind). 
2. Is it possible to support properties without creating something like `ScriptObjectWrapper`? One thing that came to mind while reviewing this PR is that we *might* be able to create property objects in C++ using pybind's C++ API for invoking Python functions (e.g. `py::module("...").attr("....")()`). So instead of creating a Python wrapper that translates `ClassType::Property` instances into Python properties with the right `self`, we may be able to do that do that in C++ in the pybind binding for `ScriptObject.getattr`. Does that make sense? Can you try that?

tugsbayasgalan commented on test/cpp/jit/test_custom_class_registrations.cpp:
```diff
@@ -28,9 +28,7 @@ struct Foo : torch::CustomClassHolder {
-  ~Foo() {
-    // std::cout<<"Destroying object with values: "<<x<<' '<<y<<std::endl;
-  }
+  ~Foo(){};
```
sounds good.

tugsbayasgalan commented on torch/custom_class.h:
```diff
@@ -177,6 +177,80 @@ class class_ {
+  template <typename T>
+  class_& def_readonly(std::string name, T CurClass::*field) {
+    auto getter_name = name + "_getter";
+    auto getter =
+        [field = std::move(field)](const c10::intrusive_ptr<CurClass>& self) {
+          return self.get()->*field;
+        };
+
+    auto wrapped_getter = detail::wrap_func<CurClass>(std::move(getter));
+    auto property_getter =
+        process_and_register_property_method(getter_name, wrapped_getter);
+
+    torch::jit::Function* property_setter;
+
+    classTypePtr->addProperty(name, property_getter, property_setter);
+    return *this;
```
sure yeah

gmagogsfm reviewed (commented):
> I think the basics are correct, but some cleanup is required.
> 
> In addition, the introduction of `ScriptObjectWrapper` is a lot messier than I thought it would be and so I have two questions:
> 
> 1. Do we really need to enable using properties on TorchBind objects outside of TorchScript (cc: @gmagogsfm) ? As I see it, Torchbind is a solution for FFI between TorchScript and C++, not Python and C++ (that's pybind).

Torchbind is indeed originally supposed to be FFI for TorchScript and C++, but TorchScript is constructed by Python, there is a real way of escaping supporting Pybind-like features. 

Unless we recommend users to register a class both for pybind and torchbind, and use them in eager and scripting mode respectively, but that requires two implementations. Take properties for example, in Pybind they can use normal `object.attr` syntax but in Torchbind they would need `object.get_attr()`. What do you think?

tugsbayasgalan commented on torch/csrc/jit/python/pybind_utils.cpp:
```diff
@@ -156,6 +156,7 @@ IValue toIValue(py::handle obj, const TypePtr& type, c10::optional<int32_t> N) {
         // if obj is already a ScriptModule, just return its ivalue
         return mod.value()._ivalue();
       }
+
```
ok

tugsbayasgalan commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -733,6 +744,7 @@ inline py::object toPyObject(IValue ivalue) {
       py::setattr(pyObj, attrName.c_str(), toPyObject(std::move(v)));
     }
     return pyObj;
+
```
ok

SplitInfinity commented:
> Torchbind is indeed originally supposed to be FFI for TorchScript and C++, but TorchScript is constructed by Python, there is a real way of escaping supporting Pybind-like features.

> Unless we recommend users to register a class both for pybind and torchbind, and use them in eager and scripting mode respectively, but that requires two implementations. Take properties for example, in Pybind they can use normal object.attr syntax but in Torchbind they would need object.get_attr(). What do you think?

Ok fair enough, but I think we should think more carefully about adding this `ScriptClassWrapper` and should even do it in a follow-up PR. That is, we land property support for use cases within TorchScript for now (that is the use case I have in mind, cc: @nikithamalgifb), and then we land property support for use cases outside of TorchScript separately.

It would be great if we could expose properties in Python without introducing `ScriptClassWrapper`; I suggested a potential way of doing that in my previous comment but I don't know for sure if it will work or not.

tugsbayasgalan commented on torch/custom_class.h:
```diff
@@ -177,6 +177,80 @@ class class_ {
+
+  template <typename T>
+  class_& def_readwrite(std::string name, T CurClass::*field) {
+    auto getterName = name + "_getter";
+    auto getter =
+        [field = std::move(field)](const c10::intrusive_ptr<CurClass>& self) {
+          return self.get()->*field;
+        };
+
+    auto wrapped_getter = detail::wrap_func<CurClass>(std::move(getter));
+    auto property_getter =
+        process_and_register_property_method(getterName, wrapped_getter);
+
+    auto setter_name = name + "_setter";
+    auto setter = [field = std::move(field)](
+                      const c10::intrusive_ptr<CurClass>& self, T value) {
+      self.get()->*field = value;
+    };
+
+    auto wrapped_setter = detail::wrap_func<CurClass>(std::move(setter));
+    auto property_setter =
+        process_and_register_property_method(setter_name, wrapped_setter);
+
+    classTypePtr->addProperty(name, property_getter, property_setter);
+
+    return *this;
```
yep sounds good

tugsbayasgalan commented on torch/custom_class.h:
```diff
@@ -286,6 +359,34 @@ class class_ {
 
+  template <typename Func>
+  torch::jit::Function* process_and_register_property_method(
+      std::string name,
```
will do

tugsbayasgalan commented on test/jit/test_torchbind.py:
```diff
@@ -130,6 +130,59 @@ def foo():
+            fooGetterSetter.x = old + 4
+            new = fooGetterSetter.x
+            return fooGetterSetter, old, new
+        scripted = torch.jit.script(foo_getter_setter_full)
+        out, old, new = scripted()
+        self.assertEqual(old, 7)
+        self.assertEqual(new, 15)
+        # check if we can use scripted Torchbind Class
+        self.assertEqual(out.x, 15)
```
For this to work, we need to define equal method ?

SplitInfinity replied:
Why? Can't you just return `old, new` as a tuple from TorchScript without returning the object?

tugsbayasgalan commented on test/jit/test_torchbind.py:
```diff
@@ -130,6 +130,59 @@ def foo():
+        self.assertEqual(out.x, 15)
+
+        def foo_just_getter():
+            fooGetterSetter = torch.classes._TorchScriptTesting._FooGetterSetter(5, 6)
+            # getY method intentionally adds 4 to x
+            return fooGetterSetter, fooGetterSetter.y
```
Updated with error messages

tugsbayasgalan commented on test/jit/test_torchbind.py:
```diff
@@ -130,6 +130,59 @@ def foo():
+
+        def foo_readwrite():
+            fooReadWrite = torch.classes._TorchScriptTesting._FooReadWrite(5, 6)
+            old = fooReadWrite.x
+            fooReadWrite.x = old + 4
+            return fooReadWrite, fooReadWrite.x, fooReadWrite.y
```
Updated with error messages

tugsbayasgalan commented on torch/csrc/jit/api/object.h:
```diff
@@ -102,6 +102,17 @@ struct TORCH_API Object {
     });
   }
 
+  using Property = std::tuple<std::string, Method, c10::optional<Method>>;
```
yep

tugsbayasgalan committed:
Update on "Add property binding in torchbind"


Summary: This PR adds property support to Torchbind. There are two cases that it needs to work:

**Torchscript**
Inside Torchscript, we don't go through pybind so there is no issue with accessing properties through ClassType.

**Eager Mode**
In Eager Mode, Torchbind creates ScriptObject which we cannot dynamically add (aka access) properties after initializing it. (https://stackoverflow.com/questions/1325673/how-to-add-property-to-a-class-dynamically
) Therefore we created a Python wrapper (ScriptObjectWrapper) around ScriptObject where we can use property method to set properties.  By doing so, we can look up wrapped object's property through __getattr__ method of the ScriptObjectWrapper. This logic is inspired from https://github.com/pytorch/pytorch/pull/44324

Test Plan: test cases in test_torchbind.py

Reviewers: @gmagogsfm, @SplitInfinity 

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan commented on torch/custom_class.h:
```diff
@@ -1,6 +1,5 @@
 #pragma once
 
-#include <ATen/core/stack.h>
```
this is what clang-format does. Should i keep it?

SplitInfinity replied:
I would say remove it.

tugsbayasgalan commented on torch/custom_class.h:
```diff
@@ -11,8 +10,9 @@
 #include <c10/util/TypeTraits.h>
-#include <torch/library.h>
 #include <torch/custom_class_detail.h>
+#include <torch/library.h>
```
This is what clang-format does. Should I keep it?

SplitInfinity replied:
I would remove it.

tugsbayasgalan committed:
Update on "Add property binding in torchbind"


Summary: This PR adds property support to Torchbind. There are two cases that it needs to work:

**Torchscript**
Inside Torchscript, we don't go through pybind so there is no issue with accessing properties through ClassType.

**Eager Mode**
In Eager Mode, Torchbind creates ScriptObject which we cannot dynamically add (aka access) properties after initializing it. (https://stackoverflow.com/questions/1325673/how-to-add-property-to-a-class-dynamically
) Therefore we created a Python wrapper (ScriptObjectWrapper) around ScriptObject where we can use property method to set properties.  By doing so, we can look up wrapped object's property through __getattr__ method of the ScriptObjectWrapper. This logic is inspired from https://github.com/pytorch/pytorch/pull/44324

Test Plan: test cases in test_torchbind.py

Reviewers: @gmagogsfm, @SplitInfinity 

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add property binding in torchbind"


Summary: This PR adds property support to Torchbind. There are two cases that it needs to work:

**Torchscript**
Inside Torchscript, we don't go through pybind so there is no issue with accessing properties through ClassType.

**Eager Mode**
In Eager Mode, Torchbind creates ScriptObject which we cannot dynamically add (aka access) properties after initializing it. (https://stackoverflow.com/questions/1325673/how-to-add-property-to-a-class-dynamically
) Therefore we created a Python wrapper (ScriptObjectWrapper) around ScriptObject where we can use property method to set properties.  By doing so, we can look up wrapped object's property through __getattr__ method of the ScriptObjectWrapper. This logic is inspired from https://github.com/pytorch/pytorch/pull/44324

Test Plan: test cases in test_torchbind.py

Reviewers: @gmagogsfm, @SplitInfinity 

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add property binding in torchbind"


Summary: This PR adds property support to Torchbind. There are two cases that it needs to work:

**Torchscript**
Inside Torchscript, we don't go through pybind so there is no issue with accessing properties through ClassType.

**Eager Mode**
In Eager Mode, Torchbind creates ScriptObject which we cannot dynamically add (aka access) properties after initializing it. (https://stackoverflow.com/questions/1325673/how-to-add-property-to-a-class-dynamically
) Therefore we created a Python wrapper (ScriptObjectWrapper) around ScriptObject where we can use property method to set properties.  By doing so, we can look up wrapped object's property through __getattr__ method of the ScriptObjectWrapper. This logic is inspired from https://github.com/pytorch/pytorch/pull/44324

Test Plan: test cases in test_torchbind.py

Reviewers: @gmagogsfm, @SplitInfinity 

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

SplitInfinity commented on test/cpp/jit/test_custom_class_registrations.cpp:
```diff
@@ -51,9 +51,6 @@ struct Foo : torch::CustomClassHolder {
-  ~Foo() {
-    // std::cout<<"Destroying object with values: "<<x<<' '<<y<<std::endl;
-  }
 };
```
Intentional?

tugsbayasgalan replied:
Yeah. this is commented out anyways so i thought i might as well delete it.

SplitInfinity replied:
Hmm, ok.

SplitInfinity commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -720,6 +720,7 @@ inline py::object toPyObject(IValue ivalue) {
     if (obj->name().find("__torch__.torch.classes") == 0) {
       return py::cast(Object(obj));
     }
+
```
Undo

SplitInfinity commented on torch/csrc/jit/api/object.h:
```diff
@@ -102,6 +102,38 @@ struct TORCH_API Object {
+  bool has_property(const std::string& name) const {
+    auto props = get_properties();
+    for (auto prop : props) {
+      if (std::get<0>(prop) == name) {
```
Consider returning a `struct` from `get_properties` so you can do `prop.name` here instead. That would be more readable.

SplitInfinity commented on torch/csrc/jit/api/object.h:
```diff
@@ -102,6 +102,38 @@ struct TORCH_API Object {
+  const Property get_property(const std::string& name) const {
+    auto props = get_properties();
+    for (auto prop : props) {
+      if (std::get<0>(prop) == name) {
```
Consider returning a `struct` from `get_properties` so you can do `prop.name` here instead. That would be more readable.

SplitInfinity commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -783,11 +786,55 @@ void initJitScriptBindings(PyObject* module) {
+                        py::module::import("builtins").attr("property");
+                    py::object property =
+                        property_func(getter_func, setter_func);
+                    return property.attr("fget")();
```
Calling from C++ into Python into C++, amazing 😍 

I know I suggested the current implementation, but looking after looking at this, we don't need to construct the property object at all. Why don't we just call `getter_func` directly? It's a `Method`, right?

SplitInfinity commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -783,11 +786,55 @@ void initJitScriptBindings(PyObject* module) {
+                        py::module::import("builtins").attr("property");
+                    py::object property =
+                        property_func(getter_func, setter_func);
+                    property.attr("fset")(value);
```
Same here.

SplitInfinity commented on torch/custom_class.h:
```diff
@@ -230,6 +230,75 @@ class class_ {
+
+  /// Property registration API for properties with only getter function.
+  template <typename GetterFunc>
+  class_& def_property(
```
All of these look much cleaner now, thanks!

SplitInfinity commented on torch/jit/_script.py:
```diff
@@ -72,7 +72,6 @@ def _compile_and_register_class(obj, rcb, qualified_name):
     torch._C._jit_script_class_compile(qualified_name, ast, defaults, rcb)
     torch.jit._state._add_script_class(obj, qualified_name)
 
-
```
Undo

SplitInfinity commented on torch/jit/_script.py:
```diff
@@ -935,7 +934,6 @@ def forward(self, input):
         warnings.warn(
             "`optimize` is deprecated and has no effect. Use `with torch.jit.optimized_execution() instead"
         )
-
```
Undo

SplitInfinity commented on torch/jit/_script.py:
```diff
@@ -999,7 +997,6 @@ def forward(self, input):
         _set_jit_function_cache(obj, fn)
         return fn
 
-
```
Undo

SplitInfinity commented on torch/jit/_trace.py:
```diff
@@ -547,7 +547,6 @@ def make_module(mod, _module_class, _compilation_unit):
     if isinstance(mod, ScriptModule):
         return mod
     elif torch._jit_internal.module_has_exports(mod):
-
```
Undo

SplitInfinity reviewed (commented):
I realized we might not even need to create the `property` objects like I suggested since returning them didn't work.

tugsbayasgalan committed:
Update on "Add property binding in torchbind"


Summary: This PR adds property support to Torchbind. There are two cases that it needs to work:

**Torchscript**
Inside Torchscript, we don't go through pybind so there is no issue with accessing properties through ClassType.

**Eager Mode**
In Eager Mode, Torchbind creates ScriptObject which we cannot dynamically add (aka access) properties after initializing it. (https://stackoverflow.com/questions/1325673/how-to-add-property-to-a-class-dynamically
) Therefore we created a Python wrapper (ScriptObjectWrapper) around ScriptObject where we can use property method to set properties.  By doing so, we can look up wrapped object's property through __getattr__ method of the ScriptObjectWrapper. This logic is inspired from https://github.com/pytorch/pytorch/pull/44324

Test Plan: test cases in test_torchbind.py

Reviewers: @gmagogsfm, @SplitInfinity 

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

SplitInfinity commented on test/jit/test_torchbind.py:
```diff
@@ -130,6 +130,74 @@ def foo():
+            return old, new
+        self.checkScript(foo_getter_setter_full, ())
+
+        def foo_getter_setter_script_check_outside():
+            fooGetterSetter = torch.classes._TorchScriptTesting._FooGetterSetter(5, 6)
+            return fooGetterSetter
+        scripted_foo_getter_setter = torch.jit.script(foo_getter_setter_script_check_outside)
+        value = scripted_foo_getter_setter()
+        self.assertEqual(value.x, 7)
+        value.x = 9
+        self.assertEqual(value.x, 13)
```
Doesn't `checkScript` test this? It tests Python + both script frontends?

tugsbayasgalan replied:
ohh that's true. Will update

tugsbayasgalan replied:
just removed it because the test above already tests this behavior

SplitInfinity commented on test/jit/test_torchbind.py:
```diff
@@ -130,6 +130,74 @@ def foo():
+        value.x = 9
+        self.assertEqual(value.x, 13)
+
+        fooGetterSetter = torch.classes._TorchScriptTesting._FooGetterSetter(1, 2)
+        self.assertEqual(fooGetterSetter.x, 3)
+        fooGetterSetter.x = 5
+        self.assertEqual(fooGetterSetter.x, 9)
```
Same here.

tugsbayasgalan replied:
done

SplitInfinity commented on test/jit/test_torchbind.py:
```diff
@@ -130,6 +130,74 @@ def foo():
+            return fooGetterSetter, fooGetterSetter.y
+        scripted = torch.jit.script(foo_just_getter)
+        out, result = scripted()
+        self.assertEqual(result, 10)
```
```suggestion
        self.assertEqual(result, 10)

```

tugsbayasgalan replied:
done

SplitInfinity commented on test/jit/test_torchbind.py:
```diff
@@ -130,6 +130,74 @@ def foo():
+            old = fooGetterSetter.y
+            fooGetterSetter.y = old + 4
+            # getY method intentionally adds 4 to x
+            return fooGetterSetter.y
```
```suggestion
            return fooGetterSetter.y

```

tugsbayasgalan replied:
done

SplitInfinity commented on test/jit/test_torchbind.py:
```diff
@@ -130,6 +130,74 @@ def foo():
+            fooReadWrite = torch.classes._TorchScriptTesting._FooReadWrite(5, 6)
+            old = fooReadWrite.x
+            fooReadWrite.x = old + 4
+            return fooReadWrite.x, fooReadWrite.y
```
```suggestion
            return fooReadWrite.x, fooReadWrite.y

```

tugsbayasgalan replied:
done

SplitInfinity commented on test/jit/test_torchbind.py:
```diff
@@ -130,6 +130,74 @@ def foo():
+        def foo_readwrite_error():
+            fooReadWrite = torch.classes._TorchScriptTesting._FooReadWrite(5, 6)
+            fooReadWrite.y = 5
+            return fooReadWrite
```
```suggestion
            return fooReadWrite

```

tugsbayasgalan replied:
done

SplitInfinity commented on torch/csrc/jit/api/object.h:
```diff
@@ -2,11 +2,18 @@
 namespace torch {
 namespace jit {
 
+struct ObjectProperty {
+  std::string name;
+  Method getter_func;
+  c10::optional<Method> setter_func;
+};
```
Consider putting this definition inside `Object`; it's not really useful as a class in itself.

tugsbayasgalan replied:
sure

SplitInfinity commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -1,5 +1,6 @@
 #include <torch/csrc/jit/python/script_init.h>
 
+#include <python3.8/descrobject.h>
```
What is this header for?

SplitInfinity commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -32,6 +34,7 @@
 #include <ATen/core/function_schema.h>
 #include <ATen/core/qualified_name.h>
 
+#include <Exceptions.h>
```
Why is this header needed?

SplitInfinity commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -40,6 +43,7 @@
 #include <cstddef>
 #include <memory>
 #include <sstream>
+#include <stdexcept>
```
Why is this header needed?

SplitInfinity commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -17,6 +18,7 @@
 #include <torch/csrc/jit/ir/constants.h>
 #include <torch/csrc/jit/ir/irparser.h>
 #include <torch/csrc/jit/passes/inliner.h>
+#include <torch/csrc/jit/python/module_python.h>
```
Why is this header needed?

tugsbayasgalan replied:
removed. clang-format is hard :(

SplitInfinity commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -783,11 +787,51 @@ void initJitScriptBindings(PyObject* module) {
                   }
+                  if (self.has_property(name)) {
+                    auto prop = self.get_property(name);
+                    auto getter_func = py::cast(prop.getter_func);
+                    auto setter_func = py::cast(prop.setter_func);
```
Hm, how come you are able to call `py::cast` without a template argument? What does the function do in this case?
```suggestion
                    auto getter_func = py::cast(prop.getter_func);
```

tugsbayasgalan replied:
Otherwise, you get "No matching function for call to object of type 'torch::jit::Method'" error. I think you can use py::cast without template argument to make the method callable python function or something. This pattern was used a lot in script_init.cpp.

tugsbayasgalan replied:
done

SplitInfinity commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -783,11 +787,51 @@ void initJitScriptBindings(PyObject* module) {
+                        name,
+                        "' which has value:",
+                        self.type()->getConstant(name));
+                  }
```
```suggestion

                  }
```

tugsbayasgalan replied:
done

SplitInfinity commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -783,11 +787,51 @@ void initJitScriptBindings(PyObject* module) {
+                try {
+                  if (self.has_property(name)) {
+                    auto prop = self.get_property(name);
+                    auto getter_func = py::cast(prop.getter_func);
```
```suggestion
```

tugsbayasgalan replied:
done

SplitInfinity commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -783,11 +787,51 @@ void initJitScriptBindings(PyObject* module) {
+                    auto prop = self.get_property(name);
+                    auto getter_func = py::cast(prop.getter_func);
+                    if (!prop.setter_func.has_value()) {
+                      TORCH_CHECK(
+                          false,
+                          "Can't set value to '",
+                          name,
+                          "' because setter function is not given.");
```
Let's find out what happens in regular Python when you try to use the setter of a property that doesn't have one and let's try to copy that behaviour as much as possible.

tugsbayasgalan replied:
cool. will verify

tugsbayasgalan replied:
It just says "can't set attribute", so I updated our error message so that it is same as python

SplitInfinity reviewed (commented):
Cool, just a few nits remaining!

tugsbayasgalan committed:
Update on "Add property binding in torchbind"


Summary: This PR adds property support to Torchbind. There are two cases that it needs to work:

**Torchscript**
Inside Torchscript, we don't go through pybind so there is no issue with accessing properties through ClassType.

**Eager Mode**
In Eager Mode, Torchbind creates ScriptObject which we cannot dynamically add (aka access) properties after initializing it. (https://stackoverflow.com/questions/1325673/how-to-add-property-to-a-class-dynamically
) Therefore we created a Python wrapper (ScriptObjectWrapper) around ScriptObject where we can use property method to set properties.  By doing so, we can look up wrapped object's property through __getattr__ method of the ScriptObjectWrapper. This logic is inspired from https://github.com/pytorch/pytorch/pull/44324

Test Plan: test cases in test_torchbind.py

Reviewers: @gmagogsfm, @SplitInfinity 

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

SplitInfinity commented on torch/csrc/jit/api/object.h:
```diff
@@ -102,6 +109,39 @@ struct TORCH_API Object {
+  }
+
+  const ObjectProperty get_property(const std::string& name) const {
+    auto props = get_properties();
```
Rather than create `ObjectProperty` instances that you don't need, why not use `type->properties()` and only create the one `ObjectProperty` you actually need?

Alternatively, you can make all of the `ObjectProperty` instances once and then get the one corresponding to `name`.

SplitInfinity commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -783,11 +783,46 @@ void initJitScriptBindings(PyObject* module) {
+                    if (!prop.setter_func.has_value()) {
+                      TORCH_CHECK(false, "can't set attribute");
+                    }
+                    auto setter_func = py::cast(prop.setter_func);
+                    setter_func(value);
```
Consider adding a comment explaining why `prop.setter_func` isn't called directly.

Potential follow-up: calling `invokeScriptMethodFromPython` here directly instead of using the pybind type for `Method`.

SplitInfinity commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -783,11 +783,46 @@ void initJitScriptBindings(PyObject* module) {
                   }
+                  if (self.has_property(name)) {
+                    auto prop = self.get_property(name);
+                    auto getter_func = py::cast(prop.getter_func);
+                    return getter_func();
```
Consider adding a comment explaining why `prop.setter_func` isn't called directly.

Potential follow-up: calling `invokeScriptMethodFromPython` here directly instead of using the pybind type for `Method`.

SplitInfinity commented on torch/csrc/jit/api/object.h:
```diff
@@ -34,6 +35,12 @@ struct TORCH_API Object {
     return _ivalue()->type();
   }
 
+  struct ObjectProperty {
+    std::string name;
+    Method getter_func;
+    c10::optional<Method> setter_func;
+  };
```
This no longer needs to be called `ObjectProperty` since it is inside `Object`. Also, please add documentation for what this struct is/does.
```suggestion
  struct Property {
    std::string name;
    Method getter_func;
    c10::optional<Method> setter_func;
  };
```

SplitInfinity commented on test/jit/test_torchbind.py:
```diff
@@ -130,6 +130,67 @@ def foo():
+        out, result = scripted()
+        self.assertEqual(result, 10)
+
+        with self.assertRaisesRegex(RuntimeError, 'can\'t set attribute'):
```
Consider using `self.assertRaisesRegexWithHighlight` here (and in other places where `assertRaisesRegex` has been used) to make sure the correct portion of the source code is highlighted.

SplitInfinity commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -783,11 +783,46 @@ void initJitScriptBindings(PyObject* module) {
+                  TypePtr type = self.type()->getAttribute(name);
+                  auto ivalue = toIValue(std::move(value), type);
+                  self.setattr(name, ivalue);
+
```
```suggestion
```

SplitInfinity commented on test/jit/test_torchbind.py:
```diff
@@ -130,6 +130,67 @@ def foo():
+        with self.assertRaisesRegex(RuntimeError, 'Tried to set read-only attribute: y'):
+            scripted = torch.jit.script(foo_not_setter)
+
+    def test_torchbind_def_readwrite(self):
```
```suggestion
    def test_torchbind_def_property_readwrite(self):
```

SplitInfinity reviewed (approved):
Nice! Just a handful of nits left.

Can you update the PR description so that it matches the contents of the PR? We have iterated on the design and structure a little since the PR was published.

Make sure also to address the `flake-8` errors (if legitimate) before landing!

tugsbayasgalan committed:
Update on "Add property binding in torchbind"


Summary: This PR adds property support to Torchbind. There are two cases that it needs to work:

**Torchscript**
Inside Torchscript, we don't go through pybind so there is no issue with accessing properties through ClassType.

**Eager Mode**
In Eager Mode, Torchbind creates ScriptObject which we cannot dynamically add (aka access) properties after initializing it. (https://stackoverflow.com/questions/1325673/how-to-add-property-to-a-class-dynamically
) Therefore we created a Python wrapper (ScriptObjectWrapper) around ScriptObject where we can use property method to set properties.  By doing so, we can look up wrapped object's property through __getattr__ method of the ScriptObjectWrapper. This logic is inspired from https://github.com/pytorch/pytorch/pull/44324

Test Plan: test cases in test_torchbind.py

Reviewers: @gmagogsfm, @SplitInfinity 

Subscribers:

Tasks:

Tags:

Differential Revision: [D26632781](https://our.internmc.facebook.com/intern/diff/D26632781)

[ghstack-poisoned]
